homeassistant/components/blueprint/models.py

```diff
@@ -0,0 +1,230 @@
+            )
+
+        # In future we can see if entities are correct domain, areas exist etc
+
+    @callback
+    def async_substitute(self) -> dict:
+        """Get the blueprint value with the inputs substituted."""
+        processed = placeholder.substitute(self.blueprint.data, self.inputs)
+        combined = {**self.config_with_inputs, **processed}
+        # From config_with_inputs
+        combined.pop(CONF_USE_BLUEPRINT)
+        # From blueprint
+        combined.pop(CONF_BLUEPRINT)
+        return combined
+
+
+class DomainBlueprints:
+    """Blueprints for a specific domain."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        domain: str,
+        logger: logging.Logger,
+    ) -> None:
+        """Initialize a domain blueprints instance."""
+        self.hass = hass
+        self.domain = domain
+        self.logger = logger
+        self._blueprints = {}
+        self._load_lock = asyncio.Lock()
+
+        hass.data.setdefault(DOMAIN, {})[domain] = self
+
+    @callback
+    def async_reset_cache(self) -> None:
+        """Reset the blueprint cache."""
+        self._blueprints = {}
+
+    def _load_blueprint(self, blueprint_path) -> Blueprint:
+        """Load a blueprint."""
+        try:
+            blueprint_data = yaml.load_yaml(
+                self.hass.config.path(BLUEPRINT_FOLDER, self.domain, blueprint_path)
+            )
+        except (HomeAssistantError, FileNotFoundError) as err:
+            raise FailedToLoad(self.domain, blueprint_path, err) from err
+
+        return Blueprint(
+            blueprint_data, expected_domain=self.domain, path=blueprint_path
+        )
+
+    def _load_blueprints(self) -> Dict[str, Union[Blueprint, BlueprintException]]:
+        """Load all the blueprints."""
+        blueprint_folder = pathlib.Path(
+            self.hass.config.path(BLUEPRINT_FOLDER, self.domain)
+        )
+        results = {}
+
+        for blueprint_path in blueprint_folder.glob("**/*.yaml"):
+            blueprint_path = str(blueprint_path.relative_to(blueprint_folder))
+            if self._blueprints.get(blueprint_path) is None:
+                try:
+                    self._blueprints[blueprint_path] = self._load_blueprint(
+                        blueprint_path
+                    )
+                except BlueprintException as err:
+                    self._blueprints[blueprint_path] = None
+                    results[blueprint_path] = err
+                    continue
+
+            results[blueprint_path] = self._blueprints[blueprint_path]
+
+        return results
+
+    async def async_get_blueprints(
+        self,
+    ) -> Dict[str, Union[Blueprint, BlueprintException]]:
+        """Get all the blueprints."""
+        async with self._load_lock:
+            return await self.hass.async_add_executor_job(self._load_blueprints)
+
+    async def async_get_blueprint(self, blueprint_name: str) -> Blueprint:
```

Comment:
Maybe rename this parameter to `blueprint_path` to keep it consistent?
```suggestion
    async def async_get_blueprint(self, blueprint_path: str) -> Blueprint:
```